Support int32 indices in torch.repeat_interleave

esqu1 opened the pull request:
Summary: To avoid casting a tensor to `.long()`, we introduce support for int32 in `torch.repeat_interleave`.

Differential Revision: D27478235

ezyang commented on aten/src/ATen/native/Repeat.cpp:
```diff
@@ -1,47 +1,69 @@
+    input = self.flatten();
+    dim = 0;
+  }
+  LOG(INFO) << "Done!";
```
kill this log line

ezyang commented on aten/src/ATen/native/cuda/Repeat.cu:
```diff
@@ -29,7 +31,11 @@ static void compute_cuda(int64_t *repeat_ptr, int64_t *cumsum_ptr, int64_t *resu
 
 Tensor repeat_interleave_cuda(const Tensor &repeat) {
-    return repeat_interleave_common<compute_cuda>(repeat);
+    Tensor output = at::empty({1});
```
You don't need to allocate a one element Tensor here, Tensor is default constructible. Just `Tensor output`.  However, more conventionally, we typically allocate the output tensor outside of `AT_DISPATCH_INDEX_TYPES` so that we don't uselessly codegen it multiple times, so it may be worth seeing if you can hoist it out.

ezyang replied:
The way the code is currently structured doesn't make it too easy to hoist. Consider porting this kernel to be structured, see https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ezyang commented on aten/src/ATen/native/Repeat.h:
```diff
@@ -2,25 +2,31 @@
+      repeats.scalar_type() == at::kLong || repeats.scalar_type() == at::kInt,
+      "repeats has to be Long or Int tensor");
+  TORCH_CHECK(
+      (repeats >= 0).all().item<uint8_t>(), "repeats can not be negative");
```
@ngimel Performance police, this looks like a sync to me

ezyang replied:
(this is a preexisting problem, no action needed)

ngimel replied:
True, it could be replaced by async assert (at the cost of nuking the context instead of error message), but even if this was done, there's computation of `total` below that definitely needs to synchronize to allocate the output tensor of the correct size, and I don't see a way around it.

ezyang commented on aten/src/ATen/native/Repeat.cpp:
```diff
@@ -1,47 +1,69 @@
-Tensor repeat_interleave_cpu(const Tensor &repeat) {
-    return repeat_interleave_common<compute_cpu>(repeat);
+Tensor repeat_interleave_cpu(const Tensor& repeat) {
+  Tensor output = at::empty({1});
```
Kill the alloc here too

Commit:
Support int32 indices in torch.repeat_interleave (#55102)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/55102

To avoid casting a tensor to `.long()`, we introduce support for int32 in `torch.repeat_interleave`.

Differential Revision: D27478235

fbshipit-source-id: a8caeade3beaef8718ab92c24c149a7491a57e35

ezyang reviewed (approved):
looks ok... tagging @mruberry in case there's a better way to setup this test

mruberry commented:
> looks ok... tagging @mruberry in case there's a better way to setup this test

There are a couple tweaks but it's OK. Pardon for the delay (it was a 4 day weekend).
